8291714: Implement a Multi-Reader Single-Writer mutex for Hotspot

jdksjolen opened the pull request:
May I please have a review for this PR which implements a `MRWMutex` class for Hotspot?

This PR does 3 things:

* Adds a port of ZGC's MRSW mutex (see [0]) to Hotspot
* Adds some new utilities for writing multi-threaded tests.
* Adds some tests for MRSW Mutex using these new utilities

The ticket has some comments which might be worth checking out: https://bugs.openjdk.org/browse/JDK-8291714

[0] Original source code here: https://github.com/openjdk/zgc/blob/zgc_generational/src/hotspot/share/gc/z/zJNICritical.cpp

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8291714](https://bugs.openjdk.org/browse/JDK-8291714): Implement a Multi-Reader Single-Writer mutex for Hotspot


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9838/head:pull/9838` \
`$ git checkout pull/9838`

Update a local copy of the PR: \
`$ git checkout pull/9838` \
`$ git pull https://git.openjdk.org/jdk pull/9838/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9838`

View PR using the GUI difftool: \
`$ git pr show -t 9838`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9838.diff">https://git.openjdk.org/jdk/pull/9838.diff</a>

</details>

jdksjolen commented:
/label add hotspot

jdksjolen commented on src/hotspot/share/runtime/mrswMutex.hpp:
```diff
@@ -0,0 +1,171 @@
+        // Wait until unblocked
+
+        // Do requested transition before blocking
+        ReaderTransition tbivm(JavaThread::current());
```
This shouldn't compute the thread, instead `read_lock` and `write_lock` should be given the threads as arguments. This will fail if the current thread isn't a Java one.

jdksjolen commented on src/hotspot/share/runtime/mrswMutex.hpp:
```diff
@@ -0,0 +1,171 @@
+        // Already blocked, wait until unblocked
+
+        // Do requested transition before blocking
+        WriterTransition tbivm(JavaThread::current());
```
This shouldn't compute the thread, instead `read_lock` and `write_lock` should be given the threads as arguments. This will fail if the current thread isn't a Java one.

fisk reviewed (commented):
64 bit cmpxchg isn't available for some 32 bit CPUs. Maybe it could be int32_t?

dholmes-ora commented:
> Multi-Reader Single-Writer mutex

Sorry to bikeshed but such a thing is called a ReadWriteLock or a ReadersWriterLock or something similar - can we please use familiar nomenclature here (even rwlock as per POSIX). "mutex" is a contraction of mutual-exclusion and as readers are not mutually exclusive it is an oxymoron to talk about a "multi-reader mutex". Thanks.

dholmes-ora commented on src/hotspot/share/runtime/mrswMutex.hpp:
```diff
@@ -0,0 +1,171 @@
+    }
+  };
+
+  template<typename T>
+  class Locker : public StackObj {
```
Why does this need to be a template? We only use it with PlatformMonitor.

dholmes-ora reviewed (commented):
Hi Johan,

A few initial comments/queries.

Thanks.

dholmes-ora commented:
> 64 bit cmpxchg isn't available for some 32 bit CPUs. Maybe it could be int32_t?

Right. AFAICS only ARMv6 would be affected by this - does anyone still build for that? (grep for SUPPORTS_NATIVE_CX8). But a 32-bit count should be more than adequate.

dholmes-ora commented:
Forgot to mention the code in the header file seems rather involved for a .hpp file - should these be in a .inline.hpp file instead?

Also did you establish this RW-lock will in fact be usable in this form for UL?

jdksjolen commented on src/hotspot/share/runtime/mrswMutex.hpp:
```diff
@@ -0,0 +1,171 @@
+class NoTransition;
+
+// Multi-reader single-writer lock implementation.
+// If a transition needs to occur when the reader or writer lock is taken (in
+// case the thread blocks), then specialize on the ReaderTransition and/or
+// WriterTransition parameters. This lock is unfair, high contention of readers
```
>Is this to allow use by different kinds of threads?

Yes, to avoid having hard-coded branches for each case. If it's only one branch then I think it can be inlined and we can skip the template. This lets us move the methods out of the header file, which is also nice.

jdksjolen commented:
>Also did you establish this RW-lock will in fact be usable in this form for UL?

I have rewritten the relevant code (`LogOutputList` class) using this RW-lock and it can run the VM in both async and synch mode. It passes all the tests for the class, and I've sent off a Mach5 job for it. I have read the code and it seems reasonable that this should work.

kimbarrett commented:
> > Multi-Reader Single-Writer mutex
> 
> Sorry to bikeshed but such a thing is called a ReadWriteLock or a ReadersWriterLock or something similar - can we please use familiar nomenclature here (even rwlock as per POSIX). "mutex" is a contraction of mutual-exclusion and as readers are not mutually exclusive it is an oxymoron to talk about a "multi-reader mutex". Thanks.

For what it's worth, C++ Standard Library uses "mutex" to refer to an object
that gets locked, and uses "lock" to refer to an RAII object that locks a
"mutex". So it has, for example, `std::shared_mutex`, which is
multi-reader/single-writer. The idea being that a "mutex" provides some kind
of mutual exclusion - in the case of shared_mutex readers (multiple) are
excluded from any writer, and writers are exclusive from each other. I'd
prefer that to calling these new things "locks".

kimbarrett commented:
HotSpot Mutex has a rank, used (in debug builds) to detect potential deadlock
situations.  These new mutexes don't have that feature, and don't interact
with Mutex in that way.  The rank system has pros and cons, but I'm not sure
just ignoring here is a good thing.

kimbarrett commented:
What is the source of the algorithm being used here?  Is there a correctness
proof?  How does it compare with, for example, Terekhov's algorithm?

dholmes-ora commented on src/hotspot/share/runtime/mrswMutex.hpp:
```diff
@@ -0,0 +1,171 @@
+  public:
+    Locker(T* lock)
+      : _lock(lock) {
+      if (_lock != nullptr) {
```
It would make sense in the same circumstances the MutexLocker also takes a NULL argument.

stefank replied:
I think the pattern of passing a nullptr argument to Lockers is a pattern that we reach for quite often. The benefit is that it allows for a conditional scope, without messing up the surrounding code.

Compare
```
Locker locker(should_lock ? &_lock : nullptr);
doit1();
doit2();
doit3();
```
vs
```
if (should_lock) {
  Locker locker(should_lock ? &_lock : nullptr);
  doit1();
  doit2();
  doit3();
} else {
  doit1();
  doit2();
  doit3();
}
```
or, having to go extra miles to pull out the code into helper functions:
```
void MyClass::doall() {
  doit1();
  doit2();
  doit3();
}

if (should_lock) {
  Locker locker(should_lock ? &_lock : nullptr);
  doall();
} else {
  doall();
}
```

The first version is much easier to read, IMHO. If you want to ban nullptrs in Lockers, I'd like to see an alternative that retains the crispness of the code.

coleenp replied:
You can have a blah_locked() version in these cases.  Having code called in multiple situations (some paths locked and some not) might be not great.  Some higher level of code should have done the locking or not.
That said, the sweeper has multiple of these with the Compile_lock so this can't be attempted to be squashed until the sweeper is gone.

stefank replied:
I've gotten more push-back on things called blah_locked(). The motivation was that it's not clear if it is the caller or callee who should take the lock, so that name convention isn't really helping the reader. I'm a but sympathetic to that point-of-view.

I still don't think this is something we should be squashing. Maybe it would help sway me if someone tried to convert ZActivatedArray and show what the code would look like if we couldn't use this pattern:
https://github.com/openjdk/zgc/blob/zgc_generational/src/hotspot/share/gc/z/zArray.inline.hpp

Alternatively, we could have two versions of lockers to appease both sides? One that never accepts nullptr, which would be the locker that most should be using, and one for those cases where we do need to run with and without the lock held.

coleenp replied:
I don't now if your zlock is even a standard runtime/mutex, but let's definitely not have another version of lockers.  I've never pushed back on blah_locked().  I need to see more evidence of why we need this in the code (after the sweeper is removed), before I suggest a policy change.
I have a patch that removes the NULL ObjectLocker argument. I thought ObjectLocker was the only thing that took null until yesterday.

dholmes-ora replied:
> Alternatively, we could have two versions of lockers to appease both sides?

"They who cannot remember the past are condemned to repeat it." :)

We used to have `MutexLockerEx` that allowed the NULL whereas `MutexLocker` did not. 

I have no issues with the NULL case as it is cleaner that defining locked and unlocked code paths.

kimbarrett replied:
MutexLockerEx did two things - allow a nullptr for the mutex argument, and
allow locking without a safepoint check. The vast majority of uses were for
the latter. Both of those got merged into MutexLocker. (The old MutexLocker
was deleted and MutexLockerEx was renamed to MutexLocker, with uses updated.)
Perhaps only the safepoint check control should have been added to
MutexLocker.

dholmes-ora commented:
> HotSpot Mutex has a rank,

True, but I would put the read/write lock in the same basket as Semaphore, being completely independent of Mutex/Monitor. The onus is on the user to use them with care. A shared ranking mechanism would be tricky I think, and read locks don't fit well with ranking schemes.

jdksjolen commented:
>How does it compare with, for example, Terekhov's algorithm?

I'm not familiar with this algorithm, do you have a link?

>What is the source of the algorithm being used here?

I'll have to defer to @fisk for this question.

dholmes-ora commented:
> For what it's worth, C++ Standard Library uses "mutex" to refer to an object that gets locked, and uses "lock" to refer to an RAII object that locks a "mutex".

So their "lock" would be our MutexLocker? The far more dominant terminology is that a mutex is a lock.

jdksjolen commented on src/hotspot/share/runtime/mrswMutex.hpp:
```diff
@@ -0,0 +1,171 @@
+// WriterTransition parameters. This lock is unfair, high contention of readers
+// may starve some of them.
+// Writers take precedence, blocking new readers from entering and allowing
+// current readers to proceed. The consequences are undefined if one or more
+// writer threads attempt to enter their critical region when another writer
+// thread already is attempting to do so or currently is in its critical region.
```
>Are you saying that this is truly restricted to only a single thread being allowed to use this in write mode, rather than just the normal one-writer-at-a-time? 

The comment was wrong, it's just the regular 1-writer-at-a-time.

kimbarrett commented:
> > How does it compare with, for example, Terekhov's algorithm?
> 
> I'm not familiar with this algorithm, do you have a link?

One place I found it.
https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2007/n2406.html#shared_mutex
search for "shared_mutex Reference Implementation"

What's proposed by this change has some similarities.  This moves some state manipulation out from under the lock by using atomic operations on the state. It only has one condvar, possibly  due to the limitations of our portability layer.  Terekhov's algorithm is fair in the sense that neither readers nor writers get preferential treatment; it defers to the OS to decide scheduling.

kimbarrett commented:
> > For what it's worth, C++ Standard Library uses "mutex" to refer to an object that gets locked, and uses "lock" to refer to an RAII object that locks a "mutex".
> 
> So their "lock" would be our MutexLocker? The far more dominant terminology is that a mutex is a lock.

Depends on which community.  I think the C++ convention makes a lot of sense,
esp. given the RAII approach to locking.

coleenp reviewed (commented):
I also want to echo David's bikeshedding comment.  Two things - I can't speak this name so am unlikely to recommend further use of this class if someone asks, and it looks too much like Mutex which has a lot of differences, and pretty simple, so don't make that more complicated to share functionality.
For some reason, I can't believe that we don't have this feature already in the vm, or at least, I was expecting this to be a simple wrapper over globalCounter.hpp. Or SingleWriterSynchronizer.hpp (whose name I can say).
If it's the same as the ZGC version, maybe the ZGC version should be moved or reconciled in the utilities directory.

kimbarrett commented:
> For some reason, I can't believe that we don't have this feature already in the vm, or at least, I was expecting this to be a simple wrapper over globalCounter.hpp. Or SingleWriterSynchronizer.hpp (whose name I can say). If it's the same as the ZGC version, maybe the ZGC version should be moved or reconciled in the utilities directory.

GlobalCounter and SingleWriterSynchronizer are RCU mechanisms. While there are
some similarities in use-cases, RCU is not equivalent to a shared mutex. I
think there are use-cases for shared mutex that aren't appropriate (or even
feasible?) for RCU. My impression is this came up because UL has its own
attempt at an RCU mechanism that is broken in both implementation and usage.
(That's my 2nd-hand impression; I've not really looked at it.) One solution
would be to change UL to use one of our other RCU mechanisms and fix any usage
problems. Another is to introduce a shared mutex and use that.  Generational
ZGC has a shared mutex to deal with a specific use-case, and this PR is
proposing to generalize that.

jdksjolen commented:
>What's proposed by this change has some similarities. This moves some state manipulation out from under the lock by using atomic operations on the state. It only has one condvar, possibly due to the limitations of our portability layer. Terekhov's algorithm is fair in the sense that neither readers nor writers get preferential treatment; it defers to the OS to decide scheduling.

Yeah, these are very similar solutions, seems like the major difference is the granularity of information provided by a condvar, locking instead of using atomics, and bit twiddling instead of just using the sign of a signed number.

jdksjolen commented:
I just pushed my first clean up of this code based on your comments.

Here's a short summary:

1. Rename to ReadWriteLock
2. Move to utilities
3. Remove template, use a conditional branch + lambda for ThreadBlockInVM (painful way of doing it, stefank has a good idea with the nullptr strategy)
4. Clarify some comments
5. Remove nullptr case from Locker, as it's not used in this code anyway
6. Use int32_t instead of int64_t for counter

Some of your comments are still not handled. Off the top of my head, I'm still not documenting everything that dholmes is interested in. @robehn has mentioned to me that there are issues when a safepointed thread holds a lock that the VM thread wants to hold, so that needs to be handled (he has also pointed out how to handle this).

Kim mentioned that UL might want to use GlobalCounter instead. This is viable, but contention on the global counter may lead to performance degradation for other users, AFAIK.

Finally, I will be simplifying this PR by pulling out the additions to threadHelper.inline.hpp into its own PR.

dholmes-ora commented on src/hotspot/share/utilities/readWriteLock.hpp:
```diff
@@ -0,0 +1,93 @@
+ *
+ */
+
+#ifndef SHARE_RUNTIME_MRSWMUTEX_HPP
+#define SHARE_RUNTIME_MRSWMUTEX_HPP
```
The include guards need updating to reflect the file renaming. (thanks for that)

dholmes-ora commented on src/hotspot/share/utilities/readWriteLock.hpp:
```diff
@@ -0,0 +1,93 @@
+class ReadWriteLock : public CHeapObj<mtSynchronizer> {
+private:
+  NONCOPYABLE(ReadWriteLock);
+  class NoTransition : public StackObj {
```
This is dead code now.

jdksjolen commented:
Thank you for the additional review, @dholmes-ora . I fixed those and added some more documentation.

The changes to `threadHelper.inline.hpp` is being merged through a separate patch (see JDK-8292679, https://github.com/openjdk/jdk/pull/9962), and so are not in need of review.

dholmes-ora reviewed (commented):
As a stand-alone RWL implementation with the described properties and limitations this seems okay to me.

Whether this RWL implementation will turn out to be useful for any existing subsystems without exhibiting performance pathologies or bad interactions with existing synchronization code is just something we will have to see. However, as we tend not to encourage unused code it would perhaps be better to delay integration of this until there is an actual usecase ready for it.

Thanks.

tstuefe commented:
I don't have time to review, just chiming in that I look forward to use RWL in NMT, replacing the lockless MST.

coleenp commented:
@tstuefe I have a memory of there being a lock like this in NMT, but it must have been removed.

If there's an upcoming use of this RWL lock, then we should integrate it first and not bundle it with the client code.

tstuefe commented:
> @tstuefe I have a memory of there being a lock like this in NMT, but it must have been removed.
> 

There is a lock for queries. MST is lockless for normal operations, which makes the table a bit cumbersome and not ideally sorted (bucket chains sorted reverse from what would be good). Normally this is not a big problem since those chains are short.

> If there's an upcoming use of this RWL lock, then we should integrate it first and not bundle it with the client code.

Certainly. I was not planning to do it right away, just voicing my support for this RWL idea.

I have to look again at this anyway, maybe I'm overlooking something. E.g. I'm not sure if the MST needs to be functional beyond VMs normal lifetime. Then it is better to leave it lockless.

jdksjolen commented:
As no current review comments were alive I rebased onto master and force pushed. This reduces the PR a bit, as parts of it (additions to `threadHelper.inline.hpp`) was merged in another PR and as such could be removed.

kimbarrett commented on src/hotspot/share/utilities/readWriteLock.cpp:
```diff
@@ -0,0 +1,147 @@
+    // Locked.
+    return;
+  }
+}
```
I was wondering why a lock for write doesn't hold the underlying
PlatformMonitor across the scope of the mutex region. Something like this
(completely untested)

```
void ReadWriteLock::write_lock(Thread* current) {
  _mon.lock();
  while (true) {
    const int32_t count = Atomic::load_acquire(&_count);
    if (count < 0) {
      // Some other writer is waiting for readers to complete.
      // Approx:
      // while (Atomic::load_acquire(&_count)) _mon.wait();
      await_write_unlock(current);
    } else if (Atomic::cmpxchg(&_count, count, -(count + 1)) == count) {
      // Claimed the write slot, but there might be active readers.
      if (count != 0) {
        // Approx:
        // while (Atomic::load_acquire(&_count) != -1) _mon.wait();
        await_no_active_readers(current);
      }
      return; // return still holding lock
    } // else failed to claim write ownership, so try again.
  }
}

void ReadWriteLock::write_unlock() {
  assert(Atomic::load(&_count) == -1, "invariant");
  Atomic::release_store(&_count, (int32_t)0);
  _mon.notify_all();
  _mon.unlock();
}
```

This seems simpler to me.  Am I missing something?

And looking at this also makes me wonder if `_mon` can just be a normal
HotSpot `Monitor`, which would resolve many of the questions and comments I've
made that are mostly driven by the direct use of `PlatformMonitor`.

jdksjolen replied:
This lock is essentially unbounded, are there any implications regarding stopping the VM from going into safepoints? I assume not, if we're using `Monitor`.

dholmes-ora commented:
@jdksjolen please do not ever force push in a live PR. There is no need, you can just do your merge locally and then push the merge changesets.

jdksjolen commented on src/hotspot/share/utilities/readWriteLock.hpp:
```diff
@@ -0,0 +1,98 @@
+  //   Attempts by readers to enter their critical region is blocked.
+  //
+  volatile int32_t _count;
+
```
It is! See L54.

jdksjolen commented on src/hotspot/share/utilities/readWriteLock.hpp:
```diff
@@ -0,0 +1,98 @@
+      _count(0) {
+  }
+
+  ~ReadWriteLock() {
```
Good point, made it explicit.

jdksjolen commented on src/hotspot/share/utilities/readWriteLock.cpp:
```diff
@@ -0,0 +1,147 @@
+        while (Atomic::load_acquire(&_count) < 0) {
+          _mon.wait(0);
+        }
+      };
```
Good point, as it's used twice and has a clear name and purpose it works well as a private member function. I also changed the other lambda into a member function, for symmetry.

> Consider putting the thread-type handling in the helper too, keeping this function small.

To me, it's less surprising to find such handling in the main method.

jdksjolen commented on src/hotspot/share/utilities/readWriteLock.cpp:
```diff
@@ -0,0 +1,147 @@
+      // critical region and the writer may now block readers.
+      if (count == -2) {
+        Locker locker(&_mon);
+        _mon.notify_all();
```
Yes. I propose leaving this work for a future RFE. There are many potential improvements, I want to await the need for those to implement them.

jdksjolen commented on src/hotspot/share/utilities/readWriteLock.cpp:
```diff
@@ -0,0 +1,147 @@
+#include "utilities/globalDefinitions.hpp"
+#include "utilities/readWriteLock.hpp"
+
+void ReadWriteLock::read_lock(Thread *current) {
```
Good find, we have two different styles in the header and this file too.

jdksjolen commented on src/hotspot/share/utilities/readWriteLock.cpp:
```diff
@@ -0,0 +1,147 @@
+
+void ReadWriteLock::read_unlock() {
+  for (;;) {
+    const int32_t count = Atomic::load_acquire(&_count);
```
Good find.

jdksjolen commented:
Latest addresses most of Kim's review comments. It also makes the `current` thread optional, allowing it to be null. This is mainly useful if the lock is used very early in initialization.

dholmes-ora commented on src/hotspot/share/utilities/readWriteLock.hpp:
```diff
@@ -80,19 +78,20 @@ class ReadWriteLock : public CHeapObj<mtSynchronizer> {
+  ~ReadWriteLock() = default;
 
-  void write_lock(Thread* current = Thread::current());
+  void write_lock(Thread* current = Thread::current_or_null());
```
I'm not seeing the motivation for allowing current to be null with these latest changes.

dholmes-ora replied:
To be clear, you will only ever get a null if someone passes it in explicitly as Thread::current_or_null() should never be able to return null in the contexts you would use this RWL.

jdksjolen replied:
>To be clear, you will only ever get a null if someone passes it in explicitly as Thread::current_or_null() should never be able to return null in the contexts you would use this RWL.

Context: UL uses static initialization for LogTagSets, these use the RWLock and as such our thread system isn't initialized yet. It's not strictly necessary, as we can re-write things to not require this.

tstuefe replied:
> >To be clear, you will only ever get a null if someone passes it in explicitly as Thread::current_or_null() should never be able to return null in the contexts you would use this RWL.
> 
> 
> 
> Context: UL uses static initialization for LogTagSets, these use the RWLock and as such our thread system isn't initialized yet. It's not strictly necessary, as we can re-write things to not require this.
> 
> 
> 
> Edit: Perhaps the right decision is to simply delay the initialization of LogTagSets.

The early initialization of UL at lib load time is certainly a pain. 

But there are other things too, other code running before initialisation, and then there is the case of non-attached threads and signal handling, etc.

I would be happy if this primitive was usable with Thread.current == Null. Otherwise you limit the future usefulness, and whatever code people use this lock in will also not be usable in these corner cases. 

If it does not cost too much complexity.

dholmes-ora replied:
Any use by non-attached threads is potentially problematic as they can prevent safepoints etc. I don't think we should be using a RWL in any code that might be executed by a non-attached thread. None of our locking code is async-signal-safe so using it from signal handlers should also be avoided.

Early initialization of UL should be examined more closely. Anything initialized before we have threads doesn't need locking by definition, so if we structure the code/api correctly we won't encounter locks. That may mean clearly separating initialization code from similar modification code e.g. a growable data structure will need a lock in the grow method, but we don't then use the grow method from the constructor.

But lets see what @jdksjolen comes up with here.

kimbarrett commented:
I don't expect this to be a popular opinion, but the more I think about it
the less I like the direct use of PlatformMonitor here rather than using
Monitor.  The "extra cruft" that comes with Monitor is there for reasons, and
I'm not convinced that bypassing all that is good.

jdksjolen commented:
People want:

1. Nullable threads
2. PlatformMonitor
3. Monitor

3. won't work with 1., as it requires a `Thread` thread, so that will require two different ReadWriteLocks (or method impls).

When we pick 3. we also have to make a decision on whether to support waiting without safepoints or not, making the behavior of ReadWriteLock a bit more complicated and perhaps opaque. Developers reading the code using this lock should preferably know what it does and what the implications are of using it.

The right answer is probably to make a `PlatformReadWriteLock` and a `ReadWriteLock`, going with convention here.

So there are a lot of different choices to be made here. On top of that we have the actual implementation, whose issues are yet to be revealed to us.

I'd like to quote David Holmes here:

>Whether this RWL implementation will turn out to be useful for any existing subsystems without exhibiting performance pathologies or bad interactions with existing synchronization code is just something we will have to see. However, as we tend not to encourage unused code it would perhaps be better to delay integration of this until there is an *actual usecase ready for it*.

I have a use case for this in UL. This is the use case which I will develop this patch towards. If new use cases arises, then patches can be made to this code to supply more use cases.

tstuefe commented:
> People want:
> 
> 1. Nullable threads
> 2. PlatformMonitor
> 3. Monitor
> 4. won't work with 1., as it requires a `Thread` thread, so that will require two different ReadWriteLocks (or method impls).
> 
> When we pick 3. we also have to make a decision on whether to support waiting without safepoints or not, making the behavior of ReadWriteLock a bit more complicated and perhaps opaque. Developers reading the code using this lock should preferably know what it does and what the implications are of using it.
> 
> The right answer is probably to make a `PlatformReadWriteLock` and a `ReadWriteLock`, going with convention here.
> 
> So there are a lot of different choices to be made here. On top of that we have the actual implementation, whose issues are yet to be revealed to us.
> 
> I'd like to quote David Holmes here:
> 
> > Whether this RWL implementation will turn out to be useful for any existing subsystems without exhibiting performance pathologies or bad interactions with existing synchronization code is just something we will have to see. However, as we tend not to encourage unused code it would perhaps be better to delay integration of this until there is an _actual usecase ready for it_.
> 
> I have a use case for this in UL. This is the use case which I will develop this patch towards. If new use cases arises, then patches can be made to this code to supply more use cases.

My use case is for the NMT malloc site table, which gives us the same restrictions as UL (being pre-init-able).

Note that even if you move UL initialization up, UL should still work without Thread==NULL in my opinion, since logging is one of the things that should work ideally always.

BTW, I am trying to get this: https://github.com/openjdk/jdk/pull/10178 into JDK, which may give you an easy option to test the mark with Thread==NULL in gtests.
